<a href="https://colab.research.google.com/github/comtechj/AIFFEL_quest_cr/blob/main/Main_Quest/0123_wine_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

wine = load_wine()

In [ ]:
wine_data = wine.data
print(wine_data.shape)

(178, 13)


In [ ]:
wine_label = wine.target
print(wine_label.shape)

(178,)


In [ ]:
wine_label_names = wine.target_names
print(wine_label_names)

['class_0' 'class_1' 'class_2']


In [ ]:
# wine.feature_names
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

:Number of Instances: 178
:Number of Attributes: 13 numeric, predictive attributes and the class
:Attribute Information:
    - Alcohol
    - Malic acid
    - Ash
    - Alcalinity of ash
    - Magnesium
    - Total phenols
    - Flavanoids
    - Nonflavanoid phenols
    - Proanthocyanins
    - Color intensity
    - Hue
    - OD280/OD315 of diluted wines
    - Proline
    - class:
        - class_0
        - class_1
        - class_2

:Summary Statistics:

============================= ==== ===== ======= =====
                                Min   Max   Mean     SD
============================= ==== ===== ======= =====
Alcohol:                      11.0  14.8    13.0   0.8
Malic Acid:                   0.74  5.80    2.34  1.12
Ash:                          1.36  3.23    2.36  0.27
Alcalinity of Ash:            10.6  30.0    19.5   3.3
Magnesium:                    70.0 162.0    99.7  14.3

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size = 0.2,
                                                    random_state = 5)

 ## Decision Tree 모델

In [ ]:
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state = 5)
decision_tree.fit(X_train, y_train)
tree_pred = decision_tree.predict(X_test)

# tree_pred == y_test
print(classification_report(tree_pred, y_test))

              precision    recall  f1-score   support

           0       0.76      1.00      0.87        13
           1       1.00      0.73      0.85        15
           2       0.88      0.88      0.88         8

    accuracy                           0.86        36
   macro avg       0.88      0.87      0.86        36
weighted avg       0.89      0.86      0.86        36



 ## Random Forest 모델

In [ ]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state = 15)
random_forest.fit(X_train, y_train)
forest_pred = random_forest.predict(X_test)

# forest_pred == y_test
print(classification_report(forest_pred, y_test))

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       1.00      0.92      0.96        12
           2       1.00      1.00      1.00         8

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



 ## SVM 모델

In [ ]:
from sklearn import svm

svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)

# svm_pred == y_test
print(classification_report(svm_pred, y_test))

              precision    recall  f1-score   support

           0       0.71      0.86      0.77        14
           1       0.82      0.69      0.75        13
           2       0.50      0.44      0.47         9

    accuracy                           0.69        36
   macro avg       0.67      0.66      0.66        36
weighted avg       0.69      0.69      0.69        36



 ## SGD Classifier 모델

In [ ]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
sgd_pred = sgd_model.predict(X_test)

# sgd_pred == y_test
print(classification_report(sgd_pred, y_test))

              precision    recall  f1-score   support

           0       0.65      0.92      0.76        12
           1       0.91      0.43      0.59        23
           2       0.00      0.00      0.00         1

    accuracy                           0.58        36
   macro avg       0.52      0.45      0.45        36
weighted avg       0.80      0.58      0.63        36



 ## Logistic Regression 모델

In [ ]:
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
logistic_pred = logistic_model.predict(X_test)

# logistic_pred == y_test
print(classification_report(logistic_pred, y_test))

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       1.00      0.79      0.88        14
           2       0.75      1.00      0.86         6

    accuracy                           0.92        36
   macro avg       0.90      0.93      0.90        36
weighted avg       0.93      0.92      0.92        36



/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# **`Random Forest 모델이 97%로 정확도가 가장 높다.`**
**`SGD 모델이 지나치게 낮다`**